### DISCUSSION TITLE
What's the best way of handling body formats for long text documents?

### DISCUSSION THREAD

**Discussion by mattiasgullstrand** [yippee-fun/phlex#564] [Help] [answered]
Hi! I'm planning on using Phlex to generate HTML for a pdf document. I want to be able to write a body on a format that keeps newlines and where it's possible to e.g. make single words bold without dividing up the text into different parts.

I might have missed something but my current solution involves `unsafe_raw` and ActionView's `simple_format`:

This is an example document HTML generator, which will be used as `PhlexDocument.new.call` to generate the HTML:
```ruby
class PhlexDocument < Phlex::HTML

  def template
    render Document.new do |doc|
      doc.body do
        <<~TEXT
          It should keep

          newlines

          It can be <b>bold</b>

          The end.
        TEXT
      end
    end
  end
end
```
And this is my Document class:
```ruby
class Document < Phlex::HTML
  include Phlex::DeferredRender
  include ActionView::Helpers::TextHelper

  def template
    if @body
      unsafe_raw(
        simple_format(
          @body.call
        )
      )
    end
  end

  def body(&block)
    @body = block
  end
end
```
Is there a better way to do this?

Thanks in advance!

**Comment by joeldrapper** [comment, marked as answer]
This looks reasonable to me. I wonder if `simple_format` has been misclassified as a “value helper” rather than an “output helper” in the `SimpleFormat` adapter. If `simple_format` is considered HTML safe (I assume it only allows safe HTML but will need to confirm this), we could make the `unsafe_raw` call in the adapter itself.

You may want to consider making the `body` an argument rather than a method. The interface could then be:

```ruby
render Document.new(<<~TEXT)
  Document text here.
TEXT
```

You may also want to look into the [phlex-markdown](https://github.com/phlex-ruby/phlex-markdown) gem, which I created while building the phlex documentation site. This is very much still in beta and subject to change, but it’s also incredibly simple, so there’s not too much that can go wrong. You could even just copy the implementation, which is entirely in [this one file](https://github.com/phlex-ruby/phlex-markdown/blob/main/lib/phlex/markdown.rb).

**Reply by mattiasgullstrand** [reply]
Thanks for the really quick response! Good recommendation, thanks for that. Will also definitely check out phlex-markdown 👍
